pkg/mimirpb/timeseries.go

```diff
@@ -288,6 +288,10 @@ func PreallocTimeseriesSliceFromPool() []PreallocTimeseries {
 
 // ReuseSlice puts the slice back into a sync.Pool for reuse.
 func ReuseSlice(ts []PreallocTimeseries) {
+	if ts == nil {
```

Comment:
if grpc receives a request where there is an empty list of timeseries, would it set this to `nil` or would it allocate it with a length of `0`? in the latter case we should probably change this to:

```suggestion
	if ts == nil || len(ts) == 0 {
```

Reply:
> receives a request where there is an empty list of timeseries

I don't think one can send "an empty list of timeseries" via protobuf. But this is safe change to do.

Reply:
Actually, I think we should check for `cap(ts)` instead.

Reply:
Done in https://github.com/grafana/mimir/pull/3889/commits/8918642834bbde08f159969b5676039e5e5f97b6.

Reply:
> I don't think one can send "an empty list of timeseries" via protobuf

I wasn't very clear... obviously one can send message with no timeseries. Standard protobuf unmarshalling would in such case leave the field as `nil`. Our unmarshalling always sets it to non-nil for `timeseries` field: https://github.com/grafana/mimir/blob/8918642834bbde08f159969b5676039e5e5f97b6/pkg/mimirpb/timeseries.go#L65-L69

But `EphemeralTimeseries` I haven't added such code. I think we don't need it, but when we set `EphemeralTimeseries` based on the matching rules, we should make sure to get slice from the pool.